pkg/matcher/cel.go

```diff
@@ -24,28 +24,93 @@ const (
+func evaluateBranchExpression(expression, branchKey string) bool {
+	env, err := cel.NewEnv(cel.Declarations(
+		decls.NewVar(branchKey, decls.String),
+	))
+	if err != nil {
+		return false
+	}
+
+	// Define data map with placeholder values.
+	data := map[string]interface{}{
+		branchKey: "refs/heads/*",
+	}
+
+	out, err := evaluateCELExpression(env, expression, data)
+	if err != nil {
+		return false
+	}
+	// Return evaluation result as bool.
+	if boolVal, ok := out.Value().(bool); ok {
+		return boolVal
+	}
+	return false
+}
+
+func handleBranchCondition(event *info.Event, data map[string]interface{}, splitExprData, branchKey, branchFromEvent string) {
+	// For push event the target_branch & source_branch info coming from payload have refs/heads/
+	// but user may or mayn't provide refs/heads/ info while giving target_branch or source_branch in CEL expression
+	// ex:  pipelinesascode.tekton.dev/on-cel-expression: |
+	//        event == "push" && target_branch == "main" && "frontend/***".pathChanged()
+	// This logic will handle such case.
+	if event.TriggerTarget == "push" {
+		if !strings.Contains(splitExprData, "refs/heads/") {
+			data[branchKey] = strings.TrimPrefix(branchFromEvent, "refs/heads/")
+		}
+	}
+	if evaluateBranchExpression(splitExprData, branchKey) {
+		data[branchKey] = "refs/heads/*"
+	}
+}
+
+// handleTargetAndSourceBranches ensures the matches of the target and source branch and update the map with correct values
+// It serves two primary purposes:
+//
+//  1. For push events, the target_branch & source_branch information from the payload have "refs/heads/" ,
+//     but users may or mayn't provide refs/heads/ prefix for target_branch or source_branch in the CEL expression.
+//     This logic handles such cases.
+//     For example:
+//     pipelinesascode.tekton.dev/on-cel-expression: |
+//     event == "push" && target_branch == "main" && "frontend/***".pathChanged()
+//
+//  2. For both push and pull_request events, if users specify target_branch and source_branch as "refs/heads/*",
+//     indicating acceptance of any branch, but the actual target_branch and source_branch values from the payload differ,
+//     it results in failure. Therefore, if users provide branch information as "refs/heads/*",
+//     the data map is updated accordingly as it indicates a positive scenario.
+//     For example:
+//
+//  1. pipelinesascode.tekton.dev/on-cel-expression: |
+//     ( event == "push" && target_branch == "refs/heads/*" && source_branch == "refs/heads/*" ) && "frontend/***".pathChanged()
+//
+//  2. pipelinesascode.tekton.dev/on-cel-expression: |
+//     event == "pull_request" && target_branch == "refs/heads/*" && source_branch == "refs/heads/*" && "frontend/***".pathChanged()
+
+func handleTargetAndSourceBranches(expr string, event *info.Event, data map[string]interface{}) {
+	splitFunc := func(c rune) bool {
+		return c == '&' || c == '(' || c == ')'
```

Comment:
should not it be `&&`

Reply:
No because its not necessary that cel expression will have always `(` and `)` its optional

Ex: 
```
1. 
      event == "pull_request" && target_branch == "refs/heads/*" && source_branch == "refs/heads/*" &&
      ( ".tekton/*.yaml".pathChanged() )
2.
      ( event == "pull_request" && target_branch == "refs/heads/*" && source_branch == "refs/heads/*" ) &&
       ".tekton/*.yaml".pathChanged()
```

Reply:
I am saying about this

```
c == '&&'
```

Reply:
It works with '&' as well 
because when splitting happens at every occurrence of &, and && is treated as regular characters in the resulting rune slices